### PR TITLE
hyperdrive marketplace integration & MMR canonicalization using fork-unique key

### PR DESCRIPTION
**Pull request by gitsimon**

### CODE REVIEW

File: pallets/hyperdrive-outgoing/src/tezos.rs [hunk 1]
```diff
@@ -77,17 +83,30 @@ fn assign_payload_schema() -> &'static Micheline {
     static ASSIGN_PAYLOAD_SCHEMA: OnceBox<Micheline> = OnceBox::new();
```
Comment (RomarQ) [inline pallets/hyperdrive-outgoing/src/tezos.rs:83]: I think this could be simplified by just having `fn assign_payload_schema() -> Micheline { ... }`.

Comment (gitsimon) [reply pallets/hyperdrive-outgoing/src/tezos.rs:83]: that would recreate the schema on each use, I guess the rust compiler is not memoizing the result

File: pallets/hyperdrive-outgoing/src/mmr_gadget/offchain_mmr.rs [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+            )),
+            Err(e) => Err(format!(
+                "Error retrieving last_message_excl_by_block from runtime API: {:?}",
+                e
+            )),
+        }
+    }
+
+    fn prune_branch(&mut self, _block_hash: &B::Hash) {
+        // TODO with the current limitation that reading offchain index is not permitted (only writing)
```
Comment (RomarQ) [inline pallets/hyperdrive-outgoing/src/mmr_gadget/offchain_mmr.rs:184]: Implementation missing.

File: pallets/hyperdrive-outgoing/Cargo.toml [hunk 3]
```diff
@@ -17,12 +17,14 @@ codec = { package = "parity-scale-codec", version = "3.0.0", default-features =
...
 sp-core = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.36" }
+jsonrpsee = { version = "0.16.2", features = ["client-core", "server", "macros"], optional = true }
+sp-blockchain = { git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.36", optional = true }
 
 # Acurast
 pallet-acurast = { path = "../acurast", default-features = false }
 pallet-acurast-marketplace = { path = "../marketplace", default-features = false}
 acurast-common = { path = "../acurast/common", default-features = false }
-mmr-lib = { package = "ckb-merkle-mountain-range", version = "0.5.2", default-features = false }
+mmr-lib = { package = "acurast-merkle-mountain-range",  git = "https://github.com/Acurast/merkle-mountain-range", default-features = false }
```
Comment (RomarQ) [inline pallets/hyperdrive-outgoing/Cargo.toml:27]: Is the fork necessary? If yes, I am fine with it.

File: pallets/hyperdrive-outgoing/src/lib.rs [hunk 4]
```diff
@@ -288,11 +339,13 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
     fn node_temp_offchain_key(
         pos: NodeIndex,
         parent_hash: <T as frame_system::Config>::Hash,
+        unique: HashOf<T, I>,
```
Comment (RomarQ) [inline pallets/hyperdrive-outgoing/src/lib.rs:342]: I do not see the need for the `unique` argument. The combination of `pos` and `parent_hash` should already be unique, right?

File: pallets/hyperdrive-outgoing/src/lib.rs [hunk 5]
```diff
@@ -311,9 +364,12 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
-            Self::snapshot_meta(Self::next_snapshot_number().saturating_sub(1))
-                .unwrap_or((0u32.into(), 0));
+        let (_root_hash, last_block, _last_message_excl): (_, T::BlockNumber, LeafIndex) =
+            Self::snapshot_meta(Self::next_snapshot_number().saturating_sub(1)).unwrap_or((
+                Default::default(),
+                0u32.into(),
+                0,
+            ));
 
         current_block.saturating_sub(last_block) >= T::MaximumBlocksBeforeSnapshot::get().into()
```
Comment (RomarQ) [inline pallets/hyperdrive-outgoing/src/lib.rs:374]: We should have into consideration the first snapshot, right?
```suggestion
        0u32.ne(&last_block) && current_block.saturating_sub(last_block) >= T::MaximumBlocksBeforeSnapshot::get().into()
```

Comment (gitsimon) [reply pallets/hyperdrive-outgoing/src/lib.rs]: fixed it properly respecting`first_mmr_block_number()`, now we don't create an empty initial snapshot when hyperdrive outgoing gets installed (by runtime upgrade) and also wait `MaximumBlocksBeforeSnapshot` before creating the first one.

**Review by RomarQ — APPROVED** [review]

**Pull request merged**